src/Type/Php/CallUserFuncDynamicFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\Type;
+
+class CallUserFuncDynamicFunctionReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array(
+			$functionReflection->getName(),
+			[
+				'call_user_func',
+				'call_user_func_array',
+			],
+			TRUE
+		);
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$args = $functionCall->args;
+
+		if (count($args) === 0) {
+			return $functionReflection->getVariants()[0]->getReturnType();
+		}
+
+		return $scope->getType(new FuncCall($firstArg->value, $args));
```

Comment:
This is wrong. The arguments should be shifted. Only 1-to-last should be passed as actual arguments.

Reply:
Where it matters - you might have some generic function called like this, and these arguments matter for correctly resolved type variable. There should be a test around that. See https://phpstan.org/blog/generics-in-php-using-phpdocs if you're not familiar with generics.

Reply:
Right, I think I understand. I have one question about this: what would be proper way to prepare params for `call_user_func_array` from `$args[1]`? Should I map it somehow myself or is there some nice API for it :) ?

Reply:
It makes sense to do it only if `$args[1]->value instanceof Array_`, otherwise I'd return the default return type.

The logic could look like this:

```php
$args = [];
foreach ($args[1]->value->items as $item) {
    $args[] = new Arg($item->value, $item->byRef, $item->unpack);
}
```